src/php/parser/Parser.php

```diff
@@ -0,0 +1,530 @@
+            {
+                if ($eof) {
+                    $matchResult = $this->reject($this->updateError($error, $position, new WildcardError()));
+                } else {
+                    $expected = mb_substr($this->input, $position, 1);
+                    $ord = IntlChar::ord($expected);
+
+                    $matchResult = $this->accept($position + strlen($expected), ASTs::asts(new Char($ord, $position), $asts), $error);
+                }
+
+                return $this->applyNext($continuations, $matchResult);
+            }
+            case ExpressionType::ALT:
+            {
+                $expressions = Expression::asAltExpression($expression)->getExpressions();
+
+                if ($expressions->isEmpty()) {
+                    return $this->applyNext($continuations, $this->reject($error));
+                } else {
+                    $cons = Continuations::cons(new AltContinuation($expressions->tail(), $position, $asts), $continuations);
+                    return $this->evalNext($expressions->head(), $position, $asts, $error, $cons);
+                }
+            }
+            case ExpressionType:: AND:
+            {
+                $cons = Continuations::cons(new AndContinuation($position, $asts, $error), $continuations);
+                return $this->evalNext($expression->getExpression(), $position, new ASTs(), $error, $cons);
+            }
+            case ExpressionType::NOT:
+            {
+                $cons = Continuations::cons(new NotContinuation($position, $asts, $error), $continuations);
+                return $this->evalNext($expression->getExpression(), $position, new ASTs(), $error, $cons);
+            }
+            case ExpressionType::VOID:
+            {
+                $cons = Continuations::cons(new VoidContinuation($asts), $continuations);
+                return $this->evalNext($expression->getExpression(), $position, new ASTs(), $error, $cons);
+            }
+            case ExpressionType::CHAR:
+            {
+                $expression = CharExpression::asCharExpression($expression);
+
+                if ($eof) {
+                    $matchResult = $this->reject($this->updateError($error, $position, new CharacterError($expression)));
+                } else {
+                    $expected = Expression::asCharExpression($expression)->getChar();
+                    $char = mb_substr($this->input, $position, 1);
+
+                    $matches = IntlChar::ord($expected) === IntlChar::ord($char);
+                    #printf("match(%s,%s)=%s, len=%s\n", IntlChar::chr($expected), $char, $matches, strlen($char));
+
+                    if (!$matches) {
+                        $matchResult = $this->reject($this->updateError($error, $position, new CharacterError($expression)));
+                    } else {
+                        $matchResult = $this->accept($position + strlen($char), ASTs::asts(new Char($char, $position), $asts), $error);
```

Comment:
`strlen` is the number of bytes but `$position` is the code point index

I think it's better to use byte indices throughout, because they allow the user to get the substring in `O(1)` later on.

If you use byte indices throughout, the next code point can be obtained in `O(1)` with `mb_strcut`
https://www.php.net/manual/en/function.mb-strcut.php